initial graphql guide commit (#238)
nautobot/nautobot: nautobot/docs/user-guides/graphql.md

# GraphQL User Guide

## Introduction

### What is GraphQL?

[GraphQL](https://graphql.org/) is a query language for your APIs and a runtime for fulfilling those queries with your existing data.

### How GraphQL simplifies API Interactions

When interacting with APIs, It's often necessary to build relationships between multiple models to achieve the result that is desired. Doing this typically requires multiple API calls to create the relationships. For example, lets assume that there are two devices in Nautobot. Each are assigned a site, region, roles, interfaces, and IP Addresses.

Simply querying the `/api/dcim/devices/` API route provides:
<div>
  <details>
    <summary>View API Results</summary>
```
{

>   "count": 208,

Reviewer: I'm making this suggestion based on the text above about "lets assume that there are two devices in Nautobot".

        }
      },
      "created": "2021-02-25",
      "last_updated": "2021-02-25T14:51:57.609598"
    }
  ]
}
```
  </details>
</div>

<br />
There is a lot of useful information in that API call, but there is also a lot of information that is missing; such as interfaces and ip addresses associated with the devices. There is also potentially a lot of information that isn't needed for the specific task. To retrieve the missing information, subsequent API calls would need to be performed; and those API results would need to be correlated to the correct device.

GraphQL reduces the complexity of performing multiple API calls and correlating results by empowering the user to create their own query that provides the user exactly what they want and nothing that they don't, in a single API call.

### Exploring GraphQL in Nautobot

In Nautobot, there is a link to the GraphQL web interface at the bottom right-hand side of the page. The GraphQL web interface is called GraphiQL. Navigating to the URI (`/graphql`), brings up the GraphiQL tool for creating queries. This interface is useful for exploring the possibilities of GraphQL and validating that written queries execute successfully.

> ![GraphiQL Interface](/user-guides/images/graphql/graphiql.png)

Reviewer: On #184, @jedelman8 had requested that we number the image filenames sequentially (so these would become `01-graphiql.png`, `02-graphiql-explorer.png,` `03-graphiql-explorer-device-query.png`, etc. I personally don't feel strongly on this point but it would be good to be consistent.

<img src="/user-guides/images/graphql/graphiql-explorer.png" alt="Documentation Explorer" width="400">

In the *Documentation Explorer*, search for `devices`. The results are all of the models that utilize the `devices` model.

<img src="/user-guides/images/graphql/graphiql-explorer-device-query.png" alt="Documentation Explorer: Devices" width="400">

From the `devices` query, select `devices` from `Query.devices`. This will display all of the potential query fields from devices.

<img src="/user-guides/images/graphql/graphiql-explorer-device-attributes.png" alt="Documentation Explorer: Devices Attributes" width="400">

### First Query

Now that you have a basic understanding of how to obtain information to query from the *Documentation Explorer*, let's craft a query. Earlier in the guide, a sample REST API call was performed to obtain device information. While the query had a lot of important information, it also lacked a lot of information. In this section, lets explore how to craft a GraphQL query that displays all of the information that we want.

> GraphQL queries are encapsulated in `query { }` flags. In recent iterations of GraphQL, `{ }` is also acceptable. With that in mind, lets craft our query. From the GraphiQL interfaces, lets query for all devices and display their device names. To do this, let's execute:

Reviewer: ```suggestion
GraphQL queries are encapsulated in `query { }` flags (simply `{ }` is also acceptable). With that in mind, let's craft our query from the GraphiQL interface to inspect all devices and display their device names. To do this, let's execute:
```

Reviewer 2: Proofreading king!
